File: Sources/GeminiSubtitles/GeminiProtocol.swift
import Foundation

/// Message-building and parsing helpers for the Gemini Live `BidiGenerateContent`
/// bidi stream. Ported from `gemini-live-translate-livekit/src/lib/translation-bridge.ts`.
///
/// The wire format is JSON over WebSocket. Outgoing messages are constructed as
/// `[String: Any]` and serialized with `JSONSerialization`; incoming messages
/// are parsed into dictionaries and probed for the documented top-level keys
/// (`setupComplete`, `serverContent`).
enum GeminiProtocol {

    static let model = "models/gemini-3.5-live-translate-preview"
    static let inputSampleRate = 48000
    static let audioMIME = "audio/pcm;rate=48000"

    // MARK: Outgoing

    /// `setup` message. See translation-bridge.ts:328-354.
    /// When `bilingual` is true, also enables `inputAudioTranscription` so the
    /// server returns the original (source) transcript alongside the
    /// translation.
    static func setupMessage(targetLanguage: String, bilingual: Bool = false) -> [String: Any] {
        var setup: [String: Any] = [
            "model": model,
            "outputAudioTranscription": [:] as [String: Any],
            "generationConfig": [
                "responseModalities": ["AUDIO"],
                "translationConfig": [
                    "targetLanguageCode": targetLanguage,
                    "echoTargetLanguage": true
                ] as [String: Any]
            ] as [String: Any],
            "realtimeInputConfig": [
                "automaticActivityDetection": ["disabled": false] as [String: Any]
            ] as [String: Any]
        ]
        if bilingual {
            // Returns a transcript of the user's input audio (the source
            // language) so the OSD can show original + translation.
            setup["inputAudioTranscription"] = [:] as [String: Any]
        }
        return ["setup": setup]
    }

    /// `realtimeInput` audio chunk. See translation-bridge.ts:601-608.
    static func realtimeAudioMessage(base64PCM: String) -> [String: Any] {
        return [
            "realtimeInput": [
                "audio": [
                    "mimeType": audioMIME,
                    "data": base64PCM
                ] as [String: Any]
            ] as [String: Any]
        ]
    }

    static func encodeJSON(_ message: [String: Any]) throws -> String {
        let data = try JSONSerialization.data(withJSONObject: message, options: [])
        return String(data: data, encoding: .utf8) ?? ""
    }

    // MARK: Incoming parsing

    /// Parsed transcription event from a server frame.
    struct TranscriptionEvent {
        let text: String
        let isFinal: Bool
        /// Source-language transcript when `inputAudioTranscription` is
        /// enabled in setup; nil otherwise.
        let original: String?
    }

    /// Extract the translation text and turnComplete flag from a server frame.
    /// Returns nil when the frame carries no `outputTranscription`.
    /// Mirrors translation-bridge.ts:356-418. Audio parts under
    /// `serverContent.modelTurn.parts[*].inlineData` are intentionally discarded.
    static func parseTranscription(from json: [String: Any]) -> TranscriptionEvent? {
        guard let serverContent = json["serverContent"] as? [String: Any] else { return nil }
        guard let transcription = serverContent["outputTranscription"] as? [String: Any],
              let text = transcription["text"] as? String, !text.isEmpty else {
            return nil
        }
        let turnComplete = (serverContent["turnComplete"] as? Bool) ?? false
        // Per the reference: isFinal = !turnComplete is inverted from typical
        // usage because each interim chunk is emitted with turnComplete=false
        // until the final chunk where turnComplete=true marks the segment end.
        // We follow the reference exactly: pass `!turnComplete` as "interim".
        return TranscriptionEvent(text: text, isFinal: !turnComplete, original: nil)
    }

    /// Extract the source-language (`inputTranscription`) text from a server
    /// frame, if present. Bilingual mode: Gemini sends input and output
    /// transcripts in SEPARATE frames, so the client has to buffer the latest
    /// input text and pair it with the next output frame.
    static func parseInputTranscription(from json: [String: Any]) -> String? {
        guard let serverContent = json["serverContent"] as? [String: Any],
              let inputTr = serverContent["inputTranscription"] as? [String: Any],
              let s = inputTr["text"] as? String, !s.isEmpty else {
            return nil
        }
        return s
    }

    static func isSetupComplete(_ json: [String: Any]) -> Bool {
        // Gemini sends `{"setupComplete": {}}` (empty object), not a bool.
        // Accept either form.
        let value = json["setupComplete"]
        if let b = value as? Bool { return b }
        if let dict = value as? [String: Any], dict.isEmpty { return true }
        if value is NSNull { return true }
        return false
    }

    static func isGoAway(_ json: [String: Any]) -> Bool {
        guard let serverContent = json["serverContent"] as? [String: Any] else { return false }
        return (serverContent["goAway"] as? Bool) ?? false
    }
}

File: shared/src/commonMain/kotlin/com/gordonxc/geminisubtitles/GeminiProtocol.kt
package com.gordonxc.geminisubtitles

import kotlinx.serialization.json.*

/**
 * Message-building and parsing helpers for the Gemini Live
 * `BidiGenerateContent` bidi stream.
 *
 * Ported from Swift `GeminiProtocol.swift` which was itself ported from
 * `gemini-live-translate-livekit/src/lib/translation-bridge.ts`.
 *
 * The wire format is JSON over WebSocket. We use kotlinx.serialization's
 * JsonObject builders for outgoing messages and JsonObject accessors for
 * incoming parsing.
 */
object GeminiProtocol {

    const val MODEL = "models/gemini-3.5-live-translate-preview"
    const val INPUT_SAMPLE_RATE = 48000
    const val AUDIO_MIME = "audio/pcm;rate=48000"

    // MARK: Outgoing

    /** `setup` message. See translation-bridge.ts:328-354. */
    fun setupMessage(targetLanguage: String): JsonObject = buildJsonObject {
        put("setup", buildJsonObject {
            put("model", MODEL)
            putJsonObject("outputAudioTranscription") {}
            putJsonObject("generationConfig") {
                putJsonArray("responseModalities") { add("AUDIO") }
                putJsonObject("translationConfig") {
                    put("targetLanguageCode", targetLanguage)
                    put("echoTargetLanguage", true)
                }
            }
            putJsonObject("realtimeInputConfig") {
                putJsonObject("automaticActivityDetection") {
                    put("disabled", false)
                }
            }
        })
    }

    /** `realtimeInput` audio chunk. See translation-bridge.ts:601-608. */
    fun realtimeAudioMessage(base64PCM: String): JsonObject = buildJsonObject {
        put("realtimeInput", buildJsonObject {
            putJsonObject("audio") {
                put("mimeType", AUDIO_MIME)
                put("data", base64PCM)
            }
        })
    }

    fun encodeJson(message: JsonObject): String =
        Json.encodeToString(JsonObject.serializer(), message)

    // MARK: Incoming parsing

    /** Parsed transcription event from a server frame. */
    data class TranscriptionEvent(
        val text: String,
        val isFinal: Boolean,
    )

    /**
     * Walks a server frame and returns a transcription event if one is present.
     * Mirrors translation-bridge.ts:356-418.
     */
    fun parseTranscription(json: JsonObject): TranscriptionEvent? {
        val serverContent = json["serverContent"] as? JsonObject ?: return null
        val transcription = serverContent["outputTranscription"] as? JsonObject ?: return null
        val text = transcription["text"]?.jsonPrimitive?.content ?: return null
        if (text.isEmpty()) return null

        val turnComplete = serverContent["turnComplete"]?.jsonPrimitive?.booleanOrNull ?: false
        // Per the reference: isFinal = !turnComplete
        return TranscriptionEvent(text = text, isFinal = !turnComplete)
    }

    /**
     * Gemini sends `{"setupComplete": {}}` (empty object), not a bool.
     * Accept either form: empty object, true, or null.
     */
    fun isSetupComplete(json: JsonObject): Boolean {
        val value = json["setupComplete"] ?: return false
        // setupComplete: {} (empty object) or true or null
        if (value is JsonObject && value.isEmpty()) return true
        if (value == JsonNull) return true
        if (value is JsonPrimitive) {
            return value.booleanOrNull == true || value.content.isEmpty()
        }
        return false
    }

    fun isGoAway(json: JsonObject): Boolean {
        val serverContent = json["serverContent"] as? JsonObject ?: return false
        return serverContent["goAway"]?.jsonPrimitive?.booleanOrNull ?: false
    }
}
